File: java_solutions_1/416.partition-equal-subset-sum.java
/*
 * @lc app=leetcode id=416 lang=java
 *
 * [416] Partition Equal Subset Sum
 */

class Solution {
    public boolean canPartition(int[] nums) {
        return java.util.Arrays.stream(nums).sum() % 2 != 0 ? false : java.util.stream.Stream.of(java.util.Arrays.stream(nums).sum() / 2).map(target -> java.util.Arrays.stream(nums).boxed().reduce(java.util.stream.Stream.of(0).map(_v -> new boolean[target + 1]).peek(dp -> dp[0] = true).findFirst().get(), (dp, num) -> java.util.stream.Stream.of(0).map(_v -> dp).peek(d -> java.util.stream.IntStream.iterate(target, i -> i >= num, i -> i - 1).forEach(i -> d[i] |= d[i - num])).findFirst().get(), (a, b) -> b)[target]).findFirst().get();
    }
}

File: kotlin_solutions/416.partition-equal-subset-sum.kt
/*
 * @lc app=leetcode id=416 lang=kotlin
 *
 * [416] Partition Equal Subset Sum
 */

class Solution {fun canPartition(nums: IntArray): Boolean = nums.sum().let { sum -> if (sum % 2 == 1) false else nums.fold(setOf(0)) { dp, n -> dp + dp.map { it + n } }.contains(sum / 2) }}

